File: SWEA/D4/1218. ［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기/［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기.cpp
#include <bits/stdc++.h>
using namespace std;
 
int main(int argc, char** argv)
{
    for(int test_case = 1; test_case <= 10; ++test_case)
    {
        ios::sync_with_stdio(0);
        cin.tie(0);
        int len;
        cin >> len;
        string input;
        cin >> input;
        if(len%2==1) {
            cout << "#" << test_case << " " << 0 << '\n';
            continue;
        }
        stack<char> st;
        for(char c : input) {
            if(c=='(' || c=='[' || c=='{' || c=='<') st.push(c);
            else if(c==')') {
                if(st.top()=='(') st.pop();
                else st.push(c);
            }
            else if(c==']') {
                if(st.top()=='[') st.pop();
                else st.push(c);
            }
            else if(c=='}') {
                if(st.top()=='{') st.pop();
                else st.push(c);
            }
            else {
                if(st.top()=='<') st.pop();
                else st.push(c);
            }
        }
        cout << "#" << test_case << " ";
        cout << (st.empty() ? 1 : 0) << '\n';
    }
    return 0;
}

File: SWEA/D4/1218. ［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기/［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Stack;
 
class Solution
{
    public static void main(String args[]) throws Exception
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        for(int test_case = 1; test_case <= 10; test_case++)
        {
            int len=Integer.parseInt(br.readLine());
            Stack<Character> st = new Stack<>();
            char[] input = br.readLine().toCharArray();
            if(len%2==1) {
                System.out.println("#"+test_case+" "+0);
                continue;
            }
            int idx=0;
            while(len>0) {
                if(input[idx]=='(' || input[idx]=='[' || input[idx]=='{' || input[idx]=='<') st.push(input[idx]);
                else if(input[idx]==')') {
                    if(!st.empty() && st.peek()=='(') st.pop();
                    else st.push(input[idx]);
                }
                else if(input[idx]==']') {
                    if(!st.empty() && st.peek()=='[') st.pop();
                    else st.push(input[idx]);
                }
                else if(input[idx]=='}') {
                    if(!st.empty() && st.peek()=='{') st.pop();
                    else st.push(input[idx]);
                }
                else {
                    if(!st.empty() && st.peek()=='<') st.pop();
                    else st.push(input[idx]);
                }
                len--;
                idx++;
            }
            if(!st.empty()) System.out.println("#"+test_case+" "+0);
            else System.out.println("#"+test_case+" "+1);
        }
    }
}